File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
def solution(want, number, discount):
    answer, temp_want = 0, []
    # temp_want: 원하는 물품을 개수 별로 넣은 뒤 정렬해 준 변수
    # zip: iterable한 객체를 인자로 받아 같이 순회한다.
    for w, num in zip(want, number):
        temp_want += [w] * num
    temp_want = sorted(temp_want)
    
    # temp_want 길이만큼만 보기
    for i in range(len(discount[:-len(temp_want)+1])):
        answer += int(temp_want == sorted(discount[i:i+len(temp_want)]))
        
    return answer

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        Map<String, Integer> wantMap = new HashMap<>();
        Map<String, Integer> discountMap = new HashMap<>();
        
        for (int i = 0; i < want.length; i++) {
            wantMap.put(want[i], number[i]);
        }        
        
        for (int i = 0; i < 10; i++) {
            discountMap.put(discount[i], discountMap.getOrDefault(discount[i], 0) + 1);
        }
        
        for (int i = 0; i <= discount.length - 10; i++) {
            if (wantMap.equals(discountMap)) {
                answer++;
            }
            
            if (i + 10 < discount.length) {
                String removeItem = discount[i];
                String addItem = discount[i + 10];

                discountMap.put(removeItem, discountMap.get(removeItem) - 1);
                if (discountMap.get(removeItem) == 0) {
                    discountMap.remove(removeItem);
                }

                discountMap.put(addItem, discountMap.getOrDefault(addItem, 0) + 1);
            }
        }
        
        return answer;
    }
}